File: codeforces/158B/py/158b.py
import sys

n = int(input())
line = input()
list = line.split()

groups = {"1": 0, "2": 0, "3": 0, "4": 0}
for i in range(len(list)):
    groups[list[i]] = groups[list[i]] + 1

result = groups['4'] + groups['3']
groups['1'] = max(0, groups['1'] - groups['3'])
result = result + int((groups['2'] + 1) / 2)
groups['1'] = max(0, groups['1'] - 2*(groups['2'] % 2))
result = result + int((groups['1'] + 3) / 4)

print(result)

File: codeforces/158B/go/158b.go
package main

import "fmt"

func max(a int, b int) int {
    if a > b {
        return a
    }
    return b
}

func main() {
    var n int = 0
    var groups = [5]int{0, 0, 0, 0, 0}

    fmt.Scanf("%d\n", &n)
    for i := 0; i < n; i++ {
        var g int
        fmt.Scanf("%d", &g)
        groups[g]++
    }

    result := groups[4] + groups[3]
    groups[1] = max(0, groups[1] - groups[3])
    result += (groups[2] + 1) / 2
    groups[1] = max(0, groups[1] - groups[2] % 2 * 2)
    result += (groups[1] + 3) / 4

    fmt.Println(result)
}
